src/NServiceBus.Metrics/RawData/RingBuffer.cs

```diff
@@ -0,0 +1,106 @@
+﻿namespace NServiceBus.Metrics.RawData
+{
+    using System;
+    using System.Threading;
+
+    class RingBuffer
+    {
+        public const int Size = 4096;
+        const long SizeMask = Size - 1;
+        const long EpochMask = ~SizeMask;
+
+        long nextToWrite;
+        long nextToConsume;
+
+        Entry[] entries = new Entry[Size];
+
+        public struct Entry
+        {
+            public Entry(long ticks, long value)
+            {
+                Ticks = ticks;
+                Value = value;
+            }
+
+            public long Ticks;
+            public long Value;
+        }
+
+        public bool TryWrite(long value)
+        {
+            var readNextWrite = Volatile.Read(ref nextToWrite);
+            long index;
+
+            do
+            {
+                var consume = Volatile.Read(ref nextToConsume);
+
+                // if writers overlaps with not consumed writes, end
+                if (readNextWrite - consume >= Size)
```

Comment:
Does this throw away new data in favor of older unreported data? In the metrics context, doesn't it make sense to always write the newest data?

If so it's always Inc the index and write to it. Then we can just serialize the whole buffer (or the latest N of it) and send that as a message. If we're paying the cost of going over the wire anyway we can afford to send too much info and sort/re-organize it later when we merge it into any other data we have.